### PR TITLE
Add AMX and EDMM and fix mmap file

### PR DESCRIPTION
**Pull request by jessehui**

### CODE REVIEW

File: src/libos/src/vm/vm_area.rs [hunk 1]
```diff
@@ -542,16 +563,39 @@ impl VMArea {
...
             let vm_range_start = self.range().start();
             for range in self.pages().get_ranges(committed) {
                 let file_offset = file_offset + (range.start() - vm_range_start);
-                file.write_at(file_offset, unsafe { range.as_slice() });
+                if file_offset >= file_len {
+                    break;
+                }
+
+                let effective_mem_len = file_len - file_offset;
+                let mem_len = std::cmp::min(PAGE_SIZE, effective_mem_len);
```
[the remark this answers is not in the document]
Comment (jessehui) [reply src/libos/src/vm/vm_area.rs]: Fixed. Thank you for your review.

File: src/libos/src/vm/vm_area.rs [hunk 2]
```diff
@@ -542,16 +563,39 @@ impl VMArea {
...
             let committed = true;
             let vm_range_start = self.range().start();
             for range in self.pages().get_ranges(committed) {
                 let file_offset = file_offset + (range.start() - vm_range_start);
-                file.write_at(file_offset, unsafe { range.as_slice() });
+                if file_offset >= file_len {
+                    break;
+                }
+
+                let effective_mem_len = file_len - file_offset;
```
Comment (volcano0dr) [inline src/libos/src/vm/vm_area.rs:591]:
```
let effective_mem_len = std::cmp::min(range().size(), file_len - file_offset);
let len = file.write_at(file_offset, unsafe {&self.as_slice()[..effective_mem_len]})?;
debug_assert!(len == effective_mem_len);
```

File: src/libos/src/vm/vm_area.rs [hunk 3]
```diff
@@ -279,7 +279,28 @@ impl VMArea {
 
         if need_flush {
             let file_offset = file_offset.unwrap() + (target_range.start() - self.range.start());
-            file.unwrap().write_at(file_offset, buf);
+            let file_len = file.unwrap().metadata().unwrap().size;
+            if file_offset < file_len {
+                let effective_mem_len = std::cmp::min(target_range.size(), file_len - file_offset);
+                if target_range == self.range() {
```
Comment (volcano0dr) [inline src/libos/src/vm/vm_area.rs:285]:
```
let effective_mem_len = std::cmp::min(target_range.size(), file_len - file_offset);
let len = file
    .unwrap()
    .write_at(file_offset, &buf[..effective_mem_len])
    .expect("flush file failure");
debug_assert!(len == effective_mem_len)
```

File: src/libos/src/vm/vm_area.rs (hunk 2 above)
Comment (jessehui) [reply src/libos/src/vm/vm_area.rs]: Done.

File: src/libos/src/vm/vm_area.rs (hunk 3 above)
Comment (jessehui) [reply src/libos/src/vm/vm_area.rs]: Done. Thanks for your review.

**Review by volcano0dr — APPROVED** [review]
LGTM

**Pull request merged**